runtime/commonMain/src/kotlinx/serialization/internal/NullableSerializer.kt

```diff
@@ -1,54 +1,73 @@
 /*
  * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
  */
+@file:UseExperimental(InternalSerializationApi::class)
 
 package kotlinx.serialization.internal
 
 import kotlinx.serialization.*
-import kotlin.reflect.KClass
 
-fun <T : Any> makeNullable(actualSerializer: KSerializer<T>): KSerializer<T?> = NullableSerializer(actualSerializer)
 
-class NullableSerializer<T : Any>(public val actualSerializer: KSerializer<T>) : KSerializer<T?> {
-    private class SerialDescriptorForNullable(val original: SerialDescriptor): SerialDescriptor by original {
-        override val isNullable: Boolean
-            get() = true
-
-        override fun equals(other: Any?): Boolean {
-            if (this === other) return true
-            if (other !is SerialDescriptorForNullable) return false
-
-            if (original != other.original) return false
+/**
+ * Returns a nullable serializer for the given serializer of non-null type.
+ */
+public val <T : Any> KSerializer<T>.nullable: KSerializer<T?>
+    get() {
+        return NullableSerializer(this)
```

Comment:
One question. What happens if the receiver is already the NullableSerializer (or perhaps a custom serializer that handles nulls directly?). I don't think that nested NullableSerializers are correct for all formats (where notnullmark is actually used)

Reply:
Good point. Checking `this.descriptor.isNullable` will also help to save some allocations.

Reply:
@qwwdfsad Could you take a look at this?

Reply:
Fixed.

> I don't think that nested NullableSerializers are correct for all formats (where notnullmark is actually used)

It should be, if you know the case where this property doesn't hold, please report